src/error_handling.c

```diff
@@ -78,9 +78,92 @@ static
 bool
 __same_string(const char * str1, const char * str2, size_t count)
 {
+  assert(NULL != str1);
+  assert(NULL != str2);
   return str1 == str2 || 0 == strncmp(str1, str2, count);
 }
 
+static
+void
+__format_overwriting_error_state_message(
+  char * buffer,
+  size_t buffer_size,
+  const rcutils_error_state_t * new_error_state)
+{
+  assert(NULL != buffer);
+  assert(0 != buffer_size);
+  assert(INT64_MAX > buffer_size);
+  assert(NULL != new_error_state);
+
+  int64_t bytes_left = buffer_size;
+  do {
+    char * offset = buffer;
+    size_t written = 0;
+
+    // write the first static part of the error message
+    written = __rcutils_copy_string(offset, bytes_left,
+        "\n"
+        ">>> [rcutils|error_handling.c:" RCUTILS_STRINGIFY(__LINE__) "] rcutils_set_error_state()\n"
+        "This error state is being overwritten:\n"
+        "\n"
+        "  '");
+    offset += written;
+    bytes_left -= written;
+    if (0 >= bytes_left) {break;}
+
+    // write the old error string
+    rcutils_error_string_t old_error_string = rcutils_get_error_string();
+    written = __rcutils_copy_string(offset, sizeof(old_error_string.str), old_error_string.str);
+    offset += written;
+    bytes_left -= written;
+    if (0 >= bytes_left) {break;}
+
+    // write the middle part of the state error message
+    written = __rcutils_copy_string(offset, bytes_left,
+        "'\n"
+        "\n"
+        "with this new error message:\n"
+        "\n"
+        "  '");
+    offset += written;
+    bytes_left -= written;
+    if (0 >= bytes_left) {break;}
+
+    // format error string for new error state and write it in
+    rcutils_error_string_t new_error_string = {
+      .str = "\0"
+    };
+    __rcutils_format_error_string(&new_error_string, new_error_state);
+    written = __rcutils_copy_string(offset, sizeof(new_error_string.str), new_error_string.str);
+    offset += written;
+    bytes_left -= written;
+    if (0 >= bytes_left) {break;}
+
+    // write the last part of the state error message
+    written = __rcutils_copy_string(offset, bytes_left,
+        "'\n"
+        "\n"
+        "rcutils_reset_error() should be called after error handling to avoid this.\n"
+        "<<<\n");
+    bytes_left -= written;
+    if (bytes_left <= 0) {break;}
```

Comment:
Nit, this check is unnecessary

Reply:
Yup, I think I can commit that one and rerun the tests locally only. f81a5c8